### PR TITLE
BUG: add infeasibility checks to min_weight_full_bipartite_matching

### LINKED ISSUE #17269: Bug in scipy.sparse.csgraph.min_weight_full_bipartite_matching
min_weight_full_bipartite_matching gets stuck when running, python is not responding to SIGINT, needed to use cltr+\ to send SIGTERM.

To reproduce:
```
import pickle
from scipy.sparse.csgraph import min_weight_full_bipartite_matching
t = ... # see comment bellow
min_weight_full_bipartite_matching(t)
```

The matrix is a coo matrix, which is representing the dense matrix below, all 0s are implicit (no 0 values in coo matrix):
```
[[0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.001      0.         0.         0.         0.         0.         0.         0.         0.        ]
 [0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.001      0.         0.         0.         0.         0.         0.         0.        ]
 [0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.001      0.         0.         0.         0.         0.         0.        ]
 [0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.001      0.         0.         0.         0.         0.        ]
 [0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.001      0.         0.         0.         0.        ]
 [0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.001      0.         0.         0.        ]
 [0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.001      0.         0.        ]
 [0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.001      0.        ]
 [0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.001     ]
 [0.         0.11687445 0.         0.         0.01319788 0.07509257 0.         0.         0.         0.74228317 0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.        ]
 [0.         0.         0.         0.81087935 0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.        ]
 [0.         0.         0.         0.         0.8408466  0.         0.         0.         0.         0.01194389 0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.        ]
 [0.         0.82994211 0.         0.         0.         0.11468516 0.         0.         0.         0.11173505 0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.        ]
 [0.18796507 0.         0.04002318 0.         0.         0.         0.         0.         0.75883335 0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.        ]
 [0.         0.         0.71545464 0.         0.         0.         0.         0.         0.02748488 0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.        ]
 [0.78470564 0.         0.         0.         0.         0.         0.         0.         0.14829198 0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.        ]
 [0.         0.10870609 0.         0.         0.         0.8918677  0.         0.         0.         0.06306644 0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.        ]
 [0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.63844085 0.         0.         0.         0.         0.         0.         0.         0.         0.        ]
 [0.         0.         0.         0.         0.         0.         0.7442354  0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.        ]
 [0.         0.         0.         0.         0.         0.         0.         0.09850549 0.         0.         0.18638258 0.2769244  0.         0.         0.         0.         0.         0.         0.         0.         0.         0.        ]
 [0.         0.         0.         0.         0.         0.         0.         0.73182464 0.         0.         0.46443561 0.38589284 0.         0.         0.         0.         0.         0.         0.         0.         0.         0.        ]
 [0.29510278 0.         0.         0.         0.         0.         0.         0.         0.09666032 0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.         0.        ]]
```

### PR DESCRIPTION
**Pull request by fuglede**
The algorithm is already equipped with a handful of checks for problem infeasibility, but the tests added here show two different examples that were not caught by any of those checks. It is likely possible to dive in and determine infeasibility in `_lapjvsp` itself, as in both examples, the symptom is the algorithm hitting an infinite loop (but a different infinite loop for each case), but here we simply add the catch-all fix of checking that the problem is feasible before even trying to solve it. This does come with a performance hit, but at least the implementation of Hopcroft--Karp that we're using is fast.

#### Reference issue
closes gh-17269

### CODE REVIEW

File: scipy/sparse/csgraph/tests/test_matching.py [hunk 1]
```diff
@@ -170,6 +171,60 @@ def test_min_weight_full_matching_infeasible_problems(biadjacency_matrix):
...
+        [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.7442354, 0.0, 0.0, 0.0,
+         0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
+        [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.09850549, 0.0, 0.0, 0.18638258,
+         0.2769244, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
+        [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.73182464, 0.0, 0.0, 0.46443561,
+         0.38589284, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0],
+        [0.29510278, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.09666032, 0.0,
+         0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0]
+        ])
+    with pytest.raises(ValueError):
```
Comment (j-bowhay) [inline scipy/sparse/csgraph/tests/test_matching.py:224]:
```suggestion
    with pytest.raises(ValueError, match='no full matching exists'):
```

**Comment by fuglede** [comment]
CI test failures seem unrelated to the change.

**Review by tylerjereddy — APPROVED** [review]
When I rebased locally, I was able to confirm that the full test suite passes on x86_64 Linux, and that this PR still fixes two separate infinite loops that are reachable in SciPy code.

Combined with the fact that no old tests were changed, only new test cases added, I think it makes sense from a release management standpoint to get this in and see if anything shows up before early 2024, but it looks like a potential performance hit in the name of correctness is justified here, and if it annoys someone they could help us improve that performance with the regression tests in place to guard us.

**Comment by tylerjereddy** [comment]
Thanks @fuglede, and @rossbar for some detailed analysis in the matching issue. If either of you see issues after this merge feel free to ping me directly--better to fix things up sooner than at release time.

**Comment by fuglede** [comment]
Thanks for diving in, @tylerjereddy!

**Pull request merged**